### PR TITLE
Streaming/single-pass transducing query processor

### PR DESCRIPTION
**Pull request by camsaul**
Like #11554, but applies to all queries executed everywhere. This will fix issues #5373, #10073, #11441, #4216 and probably many others as well.

Supersedes #9702 and #11554

This PR reworks the query to processor to transduce the result rows so they can be consumed (e.g., written to JSON) immediately. This is of course both a massive performance improvement from both a memory standpoint (since only a single row needs to be in memory at any given moment) and API endpoint response time standpoint (since we can actually start streaming result rows immediately rather than having to wait for the entire result set).

I replaced our usage of `clojure.java.jdbc` with a custom low-level JDBC-based implementation that is dramatically more efficient (for example, it resolves the methods used to read columns from results once for the entire result set instead of once per row) and much more explicit in closing resources it holds on to (the `PreparedStatement` and `ResultSet` are closed immediately when the results stop being consumed, either because we've reached the end or because the query was canceled.)

The new pattern also eliminates some hacky situations where we had to perform multiple passes to preprocess a query. Only a single pass is needed and recursive preprocessing is no longer necessary. This results in dramatically less application DB load.

The asynchronous architecture is also drastically simplified, resulting in less time waiting to run queries and easier to debug code.

### Preliminary Profiling Results: 3.25x faster

How much faster is the reworked QP? I used the following snippet to run 10,000 queries in parallel on my Ryzen 9 3950X (16-core/32-thread) machine with 32 GB of RAM:

```clj
(driver/with-driver :postgres
  (u/profile "10000 queries"
    (dorun
     (pmap
      (fn [_]
        (dotimes [_ 100]
          (-> (mt/run-mbql-query :checkins) :data :rows count))) ; or :venues
      (range 100)))))
```

**Results: an ~4x increase in throughput for queries with 10k rows!** 

🏎 🏎 🏎 🏎 🏎 🏎 🏎 🏎 🏎 🏎 🏎 🏎 🏎 🏎 🏎 🏎 

<table>
<thead>
<tr>
<th>QP</th>
<th>Run 10k queries - venues Table (100 rows)</th>
<th>Avergae query duration -- venues</th>
<th>Run 10k queries -- checkins Table (1000 rows)</th>
<th>Average query duration -- checkins</th>
</tr>
</thead>
<tbody>
<td>Current impl</td>
<td>35.3 seconds</td>
<td>~3.5 ms</td>
<td>234.0 seconds</td>
<td>~23.4 ms</td>
</tr>
<td>New Transducing QP impl</td>
<td>5.9 seconds</td>
<td>~0.59 ms</td>
<td>72.0 seconds</td>
<td>~7.2 ms</td>
</tr>
</tbody>
</table>

The results should be even more dramatic for larger result sets, but I didn't have one handy. Note also that **these results only show time to complete the entire query, but do not show time until the first results start streaming, which will be dramatically faster in the new QP.**

🏎 🏎 🏎 🏎 🏎 🏎 🏎 🏎 🏎 🏎 🏎 🏎 🏎 🏎 🏎 🏎 

###### NOTE: These examples are slightly out-of-date, I'll update them in a bit

This PR unifies the patterns used by QP middleware so all middleware is async and uses the same pattern (as opposed to the current situation where it is about half sync/half async).

Query processor middleware has the signature

```clj
(defn middleware [qp]
  (fn [query xformf chans]
    (qp query xformf chans)))
```

`query` is, of course, the query; `xformf` is a function that takes results metadata, and returns a transducer used to reduce the result rows; `chans` is a map of `core.async` promise channels used for various purposes.

Preprocessing the query can be done in-line by modifying `query`, e.g.:

```clj
(defn middleware [qp]
  (fn [query xformf {:keys [raise-chan], :as chans}]
    (try
      (qp (modify-query query) xformf chans)
      (catch Throwable e
        (a/>!! raise-chan e)))))
```

Post-processing results (i.e., modifying the results metadata or the rows) can be done by applying a composing transducing functions. `xformf` is called with results metadata before reducing results like:

```clj
(xformf metadata) -> xform
```

You can compose the `xformf`s as follows:

```clj
(defn my-xform [metadata]
  (fn [rf]
    ([]        (rf))
    ([acc]     (transform-final-result (rf acc)))
    ([acc row] (rf acc (transform-row row)))))

(defn middleware [qp]
  (fn [query xformf chans]
    (qp
     query
     (fn [metadata]
       (let [metadata' (transform-metadata metadata)]
         (comp (my-xform metadata') (xformf metadata'))))
     chans)))
```

Chans are a map of `core.async` channels used for different purposes. For writing most middleware you should only need `raise-chan` or `canceled-chan`. Instead of throwing Exceptions directly, you should send them to `raise-chan`, since the query processor is asynchronous, as demonstrated in the example above. `canceled-chan` can be used to listen for a message if the query is canceled before completion. Other channels exist for listening to various stages in the QP pipeline or for the final result.

### Driver Interface Changes

`driver/execute-query` is deprecated in favor of `driver/execute-reducible-query`. `execute-reducible-query` has the signature

```clj
(driver/execute-reducible-query driver query chans return-results)
```

And should call `return-results` with initial results metadata and a object that can be reduced to get resulting rows. Because the driver effectively decorates the reduction process by managing the context in which `return-results` is called, it can hold handles to resources (e.g. connections or `java.sql.ResultSet`s) open for the duration of the reduction and close them afterward.

`return-results` has the signature:

```clj
(return-results results-metadata rows)
```

Utility function `qp.util.reducible/reducible-rows` is introduced to simplify the implementation to create reducible, streaming results.

Example impl:

```clj
(defmethod driver/execute-reducible-query :my-driver
  [_ query chans return-results]
  (with-open [results (run-query! query)]
    (return-results
     {:cols [{:name \"my_col\"}]}
     (qp.util.reducible/reducible-rows (get-row results) chans)))
```

####  Using other reducing functions

The reducing function can be passed as a parameter to the query processor which means you can use alternate reducing functions to stream results in CSV format, XLSX format, or anything you might desire -- no more need to take the "standard" result format and then transform it as with the old QP.  This functionality could also be used to improve fingerprinters and make it possible to fingerprint rows more easily with less memory usage.

### `rff`s and access to results metadata 

The reducing function (optionally) passed to the query processor is actually a function that when passed results metadata returns a reducing function. It has the signature

```clj
(rff metadata) -> rf
```

I've been using the shorthand `rff` to refer to these functions because they are functions that return reducing functions. `rff` is thus short for `reducing-function-function`.  `xformf` is similar because it is a function that returns a transducer; `xform` is already a common Clojure abbreviation for transducer functions.

This is useful if you wanted to do something like write a reducing function that returned rows as maps rather than vectors (you need the column names from the results metadata for the keys). 

Example:

```clj
(defn- maps-rff [metadata]
  (let [ks (mapv (comp keyword :name) (:cols metadata))]
    (fn
      ([] [])

      ([acc] {:data (merge metadata {:rows acc})})

      ([acc row]
       (conj acc (zipmap ks row))))))
```

#### Manging resources in the reducing fn

The reducing function can actually be a special function created with `qp.build/decorated-reducing-fn` that decorates the reduction process itself. It is thus possible to write a reducer that opens a file only if a query completes successfully, then streams results to it:

```clj
(defn- print-rows-to-writer-rff [filename]
  (qp.build/decorated-reducing-fn
   (fn [reduce-with-rff]
     (with-open [w (io/writer filename)]
       (reduce-with-rff
        (fn [metadata]
          (fn
            ([] 0)
            ([row-count] {:rows row-count})
            ([row-count row]
             (.write w (format "ROW %d -> %s\n" (inc row-count) (pr-str row)))
             (inc row-count)))))))))

(defn print-rows-to-file-example
  "Writes results to a file. Only opens file if query is successful!"
  []
  (qp/process-query
   (driver/with-driver :postgres
     {:database (mt/id)
      :type     :query
      :query    {:source-table (mt/id :venues), :limit 20}})
   (print-rows-to-writer-rff "/tmp/results.txt")))
```

### TODO:

*  About 95% of things are working, with the main Exception being caching. I'm still working out how to cache results without retaining the entire result set in memory -- I'm considering options like streaming rows to a temp file and then loading that into the DB if we decide to cache.
*  I need to tweak quite a few tests and remove a lot of code and namespaces that are no longer needed.
*  Probably need to tweak the API response logic to properly stream QP rows when they first come in.

### CODE REVIEW

**Review by sbelak — APPROVED** [review]
This is super dope. I'll try to make fingerprinting single pass, we should be able to get rid of that extra transduce on rows, and that will give me a better feel for the changes.

File: src/metabase/query_processor/middleware/add_dimension_projections.clj [hunk 1]
```diff
@@ -196,40 +197,51 @@
...
   [qp]
-  (fn [{query-type :type, :as query}]
+  (fn [{query-type :type, :as query} xform {:keys [raise-chan], :as chans}]
     (if (= query-type :native)
-      (qp query)
-      (let [[remapping-dimensions query] (add-fk-remaps query)
-            results                      (qp query)]
-        (remap-results remapping-dimensions results)))))
+      (qp query xform chans)
+      (try
```
[the remark this answers is not in the document]
Comment (camsaul) [reply src/metabase/query_processor/middleware/add_dimension_projections.clj]: You're right, I really don't think we need to. I think I went overboard with adding all of it

File: src/metabase/query_processor/middleware/limit.clj [hunk 2]
```diff
@@ -1,20 +1,47 @@
...
+(defn- add-limit [max-rows {query-type :type, :as query}]
+  (cond-> query
+    (and (= query-type :query)
+         (qputil/query-without-aggregations-or-limits? query))
+    (assoc-in [:query :limit] max-rows)))
+
+(defn- limit-xform [max-rows]
+  (fn [xf]
+    {:pre [(fn? xf)]}
+    (let [row-count (atom 0)]
```
[the remark this answers is not in the document]
Comment (camsaul) [reply src/metabase/query_processor/middleware/limit.clj]: Yeah, totally. I forgot to do it there

File: src/metabase/query_processor/middleware/add_dimension_projections.clj [hunk 3]
```diff
@@ -196,40 +197,51 @@
...
+             internal-only-cols (map :new-column internal-only-dims)]
+         ;; now using `indexed-dims` create a function that will add internal remapped values to each row in the results
+         (reset! remap-fn (make-row-map-fn internal-only-dims))
+         (xf result (-> results-metadata
+                        ;; add remapping info `:remapped_from` and `:remapped_to` to each existing `:col`
+                        (update :cols add-remapping-info remapping-dimensions internal-only-cols)
+                        ;; now add the entries for each newly added column to the end of `:cols`
+                        (update :cols concat internal-only-cols)))))
+
+      ([result results-metadata row]
```
[the remark this answers is not in the document]
Comment (camsaul) [reply src/metabase/query_processor/middleware/add_dimension_projections.clj]: Oops. I forgot to update that one. That was a different pattern I was experimenting with initially but it ended up not playing nice with existing  tooling so I switched to the `xformf` pattern

File: src/metabase/query_processor/middleware/add_dimension_projections.clj [hunk 4]
```diff
@@ -196,40 +197,51 @@
...
+      ([result]
+       (xf result))
+
+      ([result {:keys [cols], :as results-metadata}]
+       (let [ ;; hydrate Dimensions and FieldValues for all of the columns in the results, then make a map of dimension info
+             ;; for each one that is `internal` type
+             internal-only-dims
+             (->> (hydrate cols :values :dimensions)
+                  (keep-indexed col->dim-map)
+                  (filter identity))
```
[the remark this answers is not in the document]
Comment (camsaul) [reply src/metabase/query_processor/middleware/add_dimension_projections.clj]: Great question, I just moved that code so it was already like that. I'll have to read it carefully and see if we can take it out

File: src/metabase/driver/sql_jdbc/reducible_execute.clj [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+  ^{:name (format "(.getObject rs %d %s)" col-idx (.getCanonicalName klass))}
+  (fn []
+    (.getObject rs col-idx klass)))
+
+(defmethod read-column-fn [:sql-jdbc Types/TIMESTAMP]
+  [_ rs _ i]
+  (get-object-of-class-fn rs i java.time.LocalDateTime))
+
+(defn- log-readers [driver ^ResultSetMetaData rsmeta fns]
+  (doseq [^Integer i (range 1 (inc (.getColumnCount rsmeta)))]
```
[the remark this answers is not in the document]
Comment (camsaul) [reply src/metabase/driver/sql_jdbc/reducible_execute.clj]: Yeah, definitely. I need to go back in and clean some of that stuff up still. It's still in a littered with printlns to help me debug stage

File: src/metabase/query_processor/build.clj [hunk 6]
```diff
@@ -0,0 +1,283 @@
...
+             [util :as u]]
+            [metabase.query-processor.error-type :as error-type]
+            [potemkin.types :as p.types]))
+
+;; Signatures:
+;;
+;; (Driver) execute-query-reducible:
+;;
+;; (execute-query-reducible: driver query respond raise canceled-chan) -> reducible result
+;;
```
Comment (camsaul) [inline src/metabase/query_processor/build.clj]: These comments are out-if-date please ignore. They were from a different pattern I experimented with

File: src/metabase/query_processor/build.clj [hunk 7]
```diff
@@ -0,0 +1,283 @@
...
+       (try
+         (qp
+          query
+          (fn [metadata]
+            (fn [rff]
+              (rff metadata)))
+          chans)
+         (catch Throwable e
+           (a/put! raise-chan e)))))))
+
```
Comment (camsaul) [inline src/metabase/query_processor/build.clj]: Note to self: change this to `a/>!!`

File: src/metabase/query_processor/interface.clj [hunk 8]
```diff
@@ -18,3 +19,14 @@
   "Should we disable logging for the QP? (e.g., during sync we probably want to turn it off to keep logs less
   cluttered)."
   false)
+
+;; TODO - put this somewhere better
+
+(defn do-maybe-with-open [x f]
```
Comment (camsaul) [inline src/metabase/query_processor/interface.clj]: Note to self: ended up not using this. Remove from final PR

File: src/metabase/query_processor/middleware/annotate.clj [hunk 9]
```diff
@@ -467,22 +475,39 @@
...
+   (merge-cols-returned-by-driver (column-info query result) cols-returned-by-driver)))
+
+(def ^:private column-info-sample-size
+  "Number of result rows to sample when adding column info to results."
+  100)
+
+(defn- add-column-info-xform [{query-type :type, :as query} metadata]
+  (fn [rf]
+    (redux/post-complete
+     (redux/juxt rf ((take column-info-sample-size) conj))
```
[the remark this answers is not in the document]
Comment (camsaul) [reply src/metabase/query_processor/middleware/annotate.clj]: That is a great idea 👍

[the remark this answers is not in the document]
Comment (sbelak) [reply src/metabase/query_processor/middleware/annotate.clj]: I'm happy to do this if want to split the work.

[the remark this answers is not in the document]
Comment (camsaul) [reply src/metabase/query_processor/middleware/annotate.clj]: Yeah, that would be greatly appreciated. Let me get this to a point where the tests are passing first tho so you're not fighting stuff I haven't fixed yet

**Pull request merged**